registry/group_master.go

```diff
@@ -0,0 +1,170 @@
+package registry
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/mailgun/go-etcd/etcd"
+	"github.com/mailgun/scroll/vulcan"
+)
+
+const (
+	frontendKey   = "%s/frontends/%s.%s/frontend"
+	middlewareKey = "%s/frontends/%s.%s/middlewares/%s"
+	backendKey    = "%s/backends/%s/backend"
+	serverKey     = "%s/backends/%s/servers/%s"
+)
+
+/*
+GroupMasterRegistry is an implementation of Registry that uses a single master
+instance of an application within a given group to handle requests. When the
+master instance fails, request handling will automatically failover to a slave
+instance.
+*/
+type GroupMasterRegistry struct {
```

Comment:
I got it now, I think this `GroupMaster` is a bith thick base as it implements everything that both classes know. I would make it thinner or refactor it as a utility functions instead. E.g. the notion of `Master` is not really needed for the simple case, but you still introduce it because this GroupMaster keeps notion of master inside.